queue/abstract/driver/utubettl.lua

```diff
@@ -215,9 +215,36 @@ function method.normalize_task(self, task)
     return task and task:transform(i_next_event, i_data - i_next_event)
 end
 
+-- check if mvcc is enabled
+local function check_mvcc_state()
+    local cfg = box.cfg
+    if cfg['memtx_use_mvcc_engine'] == nil  or cfg['memtx_use_mvcc_engine'] == false then
+        return false
+    end
+
+    return true
+end
+
 -- put task in space
 function method.put(self, data, opts)
-    local max = self.space.index.task_id:max()
+    local max
+
+    -- taking the maximum of the index is an implicit transactions, so it is
+    -- always done with 'read-confirmed' mvcc isolation level.
+    -- It can lead to errors when trying to make parallel 'put' calls with mvcc enabled.
+    -- It is hapenning because 'max' for several puts in parallel will be the same since
+    -- read confirmed isolation level makes visible all transactions that finished the commit.
+    -- To fix it we wrap it with box.begin/commit and set right isolation level.
+    -- See https://github.com/tarantool/queue/issues/207
+    -- See https://www.tarantool.io/ru/doc/latest/concepts/atomic/txn_mode_mvcc/
+    if check_mvcc_state() then
+        box.begin({txn_isolation = 'read-committed'})
```

Comment:
Is it possible that external transaction will be opened by user? In that case you will open nested transaction. But Tarantool doesn't support it.

Reply:
discussed in a private call

Reply:
Please, summarize it for ones who were not on your call.

Reply:
Hi! The talk was discussing how an external transaction can occur and how valid it is, whether there have been use cases and how valid they are.
I have the following thoughts on this: the work is done with the top-level queue API `put()` (not with operations typical for a DB such as insert/select/...), which doesn't have any contracts on the internals, including the absence of transactions inside.